### ISSUE TITLE
Packages with bin commands fall over on windows

### ISSUE THREAD

**Issue by AlastairTaft** [lerna/lerna#385]
To start with, everything builds okay on my MAC (although I'm running node v6.2.1). But when I run `lerna bootstrap` on my windows machine (node v4.6.1) it trips up on my package that has bin commands. Here's the log

C:\projects\Redacted>lerna bootstrap
Lerna v2.0.0-beta.30
Bootstrapping 8 packages
Installing external dependencies
Symlinking packages and binaries
Symlink already exists for @luna/csv-to-json dependency of @luna/redacted-logic1, but links to di
fferent location. Replacing with updated symlink...
@luna/basic-theme                                  ╣██████░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░╠
@luna/redacted-theme                                    ╣████████████░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░╠
@luna/csv-to-json                                  ╣██████████████████░░░░░░░░░░░░░░░░░░░░░░░░░░░░░╠
@luna/redacted-logic1                           ╣████████████████████████░░░░░░░░░░░░░░░░░░░░░░░╠
Errored while running BootstrapCommand.execute
Error: ENOENT: no such file or directory, stat 'C:\projects\Redacted\packages\csv-to-json\lib
\jsonToCsv.js'
    at Error (native)

My problem is that this file it's trying to access won't be created until it runs 'prepublish' on the module, which isn't ran yet. If I create an empty JavaScript file for where it's looking it gets past without erroring and then goes on to run 'prepublish'.

**Comment by AlastairTaft** [comment]
Same error on Windows node v6.2.1

I guess this is something to do with how the bin commands are setup on Windows that differs from Mac.

**Comment by AlastairTaft** [comment]
My workaround is I've just checked in the generated code, for now.

**Comment by thejameskyle** [comment]
You are building the `bin` files as they are installed?

**Comment by AlastairTaft** [comment]
Yeah I think so

**Comment by thejameskyle** [comment]
This sounds like an issue with the project setup and npm. Can you create a repo that reproduces it for me?

**Comment by AlastairTaft** [comment]
Yeah it's likely this isn't a lerna issue and possibly an npm issue. I'll see if I can set one up. Thanks.

**Comment by Ciantic** [comment]
This could be a lerna issue, all other bin scripts in Windows looks like this:

e.g. sha.js.cmd
```
@IF EXIST "%~dp0\node.exe" (
  "%~dp0\node.exe"  "%~dp0\..\sha.js\bin.js" %*
) ELSE (
  @SETLOCAL
  @SET PATHEXT=%PATHEXT:;.JS;=;%
  node  "%~dp0\..\sha.js\bin.js" %*
)
```

but the ones Lerna has done looks like this:

e.g. my-script.js.cmd
```
@"%~dp0\..\..\..\my-app\my-script.js"   %*
```

There is no node.exe checking at all.

**Comment by kwelch** [comment]
Not sure if this is the same at all, but I am having similar issues with windows when attempting to `lerna run test`. The initial log line comes out that says the lerna version but nothing after. I had it running for 7 hours yesterday on a CI server without realizing. This worked fine on my MBP, but on the windows box I would not get it to work. Calling the scripts directly did work on the windows box

**Comment by billyzkid** [comment]
I get the same error using Lerna 2.0.0-beta.37 on Windows for the same reason stated above. Lerna fails to symlink one of my packages (with a bin script) during `npm install` because the build folder doesn't exist yet, since it doesn't get built until `prepublish` is run after install.

```sh
Lerna v2.0.0-beta.37
Independent Versioning Mode
Bootstrapping 7 packages
Preinstalling packages
Installing external dependencies
Symlinking packages and binaries
FileSystemUtilities.symlink   ("C:\\Code\\untitled\\packages\\untitled-scripts\\build\\index.js", "C:\\Code\\untitled\\packages\\ex...)
Error: ENOENT: no such file or directory, stat 'C:\Code\untitled\packages\untitled-scripts\build\index.js'
    at Error (native)
Errored while running BootstrapCommand.execute
```

And this is the relevant part of my package.json

```json
{
  "bin": {
    "untitled-scripts": "build/index.js"
  }
}

**Comment by kirikiwi** [comment]
I am still seeing this issue today as described above.

**Comment by mgrenier** [comment]
This still happens on lerna v3.4.1. Seems like `./bin/my-command` and `./bin/my-command.cmd` that are generated from lerna are missing a lot of lines :

From lerna :
```
@"%~dp0\..\..\..\..\konstellio\cli\dist\cli.js"   %*
```
Temporary fix to match what is generated from npm :
```
@IF EXIST "%~dp0\node.exe" (
  "%~dp0\node.exe"  "%~dp0\..\..\..\..\konstellio\cli\dist\cli.js" %*
) ELSE (
  @SETLOCAL
  @SET PATHEXT=%PATHEXT:;.JS;=;%
  node  "%~dp0\..\..\..\..\konstellio\cli\dist\cli.js" %*
)
```

_Note that is for the `*.cmd` file but the linux executable also shows huge difference with the npm one._

**Comment by hulucc** [comment]
The problem still happens on v3.22.1. It's 2021.

**Comment by svdoever** [comment]
I'm afraid it is not solved with v4.0.0 yet.
Unzip the following zip file: [lerna-invalid-cmd-on-windows.zip](https://github.com/lerna/lerna/files/7916551/lerna-invalid-cmd-on-windows.zip)

- Execute `lerna bootstrap`
- Open the file `lerna-invalid-cmd-on-windows\packages\consume\node_modules\.bin\test.cmd`

Result:
```
@ECHO off
GOTO start
:find_dp0
SET dp0=%~dp0
EXIT /b
:start
SETLOCAL
CALL :find_dp0
"%dp0%\..\..\..\test\test.js"   %*
```
Expected something like:

```
@ECHO off
SETLOCAL
CALL :find_dp0

IF EXIST "%dp0%\node.exe" (
  SET "_prog=%dp0%\node.exe"
) ELSE (
  SET "_prog=node"
  SET PATHEXT=%PATHEXT:;.JS;=;%
)

"%_prog%"  "%dp0%\..\test\test.js" %*
ENDLOCAL
EXIT /b %errorlevel%
:find_dp0
SET dp0=%~dp0
EXIT /b
```

Result: I can't use the `test` command in npm scripts.
